File: src/core/contextual-registry.ts
import type { ContextGraphConfig } from "../types/config.js";
import {
  DEFAULT_SIMILARITY_THRESHOLD,
  DEFAULT_VECTOR_SEARCH_LIMIT,
} from "../types/config.js";
import type {
  DecisionTrace,
  FormattedContext,
  ScoredDecisionTrace,
} from "../types/data-model.js";
import type { EmbeddingProvider } from "../embeddings/provider.js";
import type { MultiTenantGraphmindStore } from "../db/multi-tenant-store.js";
import type { RuntimeTenantContext } from "../db/multi-tenant-store.js";
import { createLogger, type Logger } from "../utils/logger.js";

export class ContextualRegistry {
  private multiTenantStore: MultiTenantGraphmindStore;
  private embeddingProvider: EmbeddingProvider;
  private config: ContextGraphConfig;
  private logger: Logger;
  private discoveryMode: Map<string, boolean> = new Map();

  constructor(
    multiTenantStore: MultiTenantGraphmindStore,
    embeddingProvider: EmbeddingProvider,
    config: ContextGraphConfig,
  ) {
    this.multiTenantStore = multiTenantStore;
    this.embeddingProvider = embeddingProvider;
    this.config = config;
    this.logger = createLogger(config.debug);
  }

  async isDiscoveryMode(
    runtimeContext?: RuntimeTenantContext,
  ): Promise<boolean> {
    const store = await this.multiTenantStore.getStoreForRuntime(runtimeContext);
    const cacheKey = store.getGraphName();
    
    const cached = this.discoveryMode.get(cacheKey);
    if (cached !== undefined) return cached;
    
    const count = await store.countTraces();
    const isDiscovery = count === 0;
    this.discoveryMode.set(cacheKey, isDiscovery);
    
    if (isDiscovery) {
      this.logger.info(
        "Discovery Mode: no prior traces found for tenant %s, project %s",
        store.getTenant(),
        store.getProject(),
      );
    }
    return isDiscovery;
  }

  async getRelevantContext(
    intentDescription: string,
    runtimeEmbeddingProvider?: EmbeddingProvider,
    runtimeContext?: RuntimeTenantContext,
  ): Promise<FormattedContext> {
    const store = await this.multiTenantStore.getStoreForRuntime(runtimeContext);
    const embeddingProvider = this.getEmbeddingProvider(
      runtimeEmbeddingProvider,
    );
    const embedding = await embeddingProvider.embed(intentDescription);
    const limit = this.config.vectorSearchLimit ?? DEFAULT_VECTOR_SEARCH_LIMIT;

    const [pastTraces, rules, antiPatterns, skills, schema] = await Promise.all(
      [
        store.findSimilarTraces(embedding, limit),
        store.getActiveRules(),
        store.getAntiPatterns(),
        store.getSkillsByProject(),
        store.getSchemaOverview(),
      ],
    );

    this.logger.debug(
      "Retrieved context for tenant %s: %d traces, %d rules, %d anti-patterns, %d skills, %d entity types",
      store.getTenant(),
      pastTraces.length,
      rules.length,
      antiPatterns.length,
      skills.length,
      schema.nodeLabels.length,
    );

    return { pastTraces, rules, antiPatterns, skills, schema };
  }

  async recordDecision(
    trace: Omit<DecisionTrace, "id" | "createdAt" | "updatedAt">,
    runtimeEmbeddingProvider?: EmbeddingProvider,
    runtimeContext?: RuntimeTenantContext,
  ): Promise<string> {
    const store = await this.multiTenantStore.getStoreForRuntime(runtimeContext);
    const embeddingProvider = this.getEmbeddingProvider(
      runtimeEmbeddingProvider,
    );
    // Generate trace embedding from combined text
    const traceText = [
      `Intent: ${trace.intent.description}`,
      ...trace.constraints.map(
        (c) => `Constraint (${c.type}): ${c.description}`,
      ),
      `Action: ${trace.action.description}`,
      `Justification: ${trace.justification.description}`,
    ].join("\n");

    const [traceEmbedding, intentEmbedding] =
      await embeddingProvider.embedBatch([traceText, trace.intent.description]);

    // Embed constraints and action
    const constraintTexts = trace.constraints.map((c) => c.description);
    const constraintEmbeddings =
      constraintTexts.length > 0
        ? await embeddingProvider.embedBatch(constraintTexts)
        : [];
    const [actionEmbedding] = await embeddingProvider.embedBatch([
      trace.action.description,
    ]);

    const enrichedTrace = {
      ...trace,
      embedding: traceEmbedding,
      intent: { ...trace.intent, embedding: intentEmbedding },
      constraints: trace.constraints.map((c, i) => ({
        ...c,
        embedding: constraintEmbeddings[i],
      })),
      action: { ...trace.action, embedding: actionEmbedding },
    };

    const traceId = await store.saveDecisionTrace(enrichedTrace);

    // Embed concept names and update concept nodes with embeddings
    if (trace.concepts && trace.concepts.length > 0) {
      try {
        const conceptEmbeddings = await embeddingProvider.embedBatch(
          trace.concepts,
        );
        for (let i = 0; i < trace.concepts.length; i++) {
          await store.ensureConcept(
            trace.concepts[i],
            undefined,
            conceptEmbeddings[i],
          );
        }
      } catch (err) {
        this.logger.debug(
          "Failed to embed concepts: %s",
          (err as Error).message,
        );
      }
    }

    // Semantic generalization: link to similar past traces
    await this.linkPrecedents(traceId, traceEmbedding, store);

    // Reset discovery mode cache for this tenant
    this.discoveryMode.delete(store.getGraphName());

    this.logger.debug("Recorded decision trace: %s for tenant: %s", traceId, store.getTenant());
    return traceId;
  }

  private async linkPrecedents(
    newTraceId: string,
    embedding: number[],
    store: Awaited<ReturnType<typeof this.multiTenantStore.getStoreForRuntime>>,
  ): Promise<void> {
    const threshold =
      this.config.similarityThreshold ?? DEFAULT_SIMILARITY_THRESHOLD;
    const similarTraces = await store.findSimilarTraces(embedding, 5);

    for (const { trace, similarity } of similarTraces) {
      if (trace.id && trace.id !== newTraceId && similarity >= threshold) {
        await store.createPrecedentLink(newTraceId, trace.id, similarity);
        this.logger.debug(
          "Linked trace %s -> %s (similarity: %.3f)",
          newTraceId,
          trace.id,
          similarity,
        );
      }
    }
  }

  private getEmbeddingProvider(
    override?: EmbeddingProvider,
  ): EmbeddingProvider {
    return override ?? this.embeddingProvider;
  }

  async findPrecedents(
    traceId: string,
    runtimeContext?: RuntimeTenantContext,
  ): Promise<ScoredDecisionTrace[]> {
    const store = await this.multiTenantStore.getStoreForRuntime(runtimeContext);
    const trace = await store.getTraceById(traceId);
    if (!trace?.embedding) return [];
    return store.findSimilarTraces(trace.embedding);
  }
}

File: python-sdk/src/graphmind_context_graphs/core/contextual_registry.py
"""ContextualRegistry — semantic retrieval and decision recording."""

from __future__ import annotations

from ..types.config import ContextGraphConfig, DEFAULT_SIMILARITY_THRESHOLD, DEFAULT_VECTOR_SEARCH_LIMIT
from ..types.data_model import DecisionTrace, FormattedContext, ScoredDecisionTrace
from ..embeddings.provider import EmbeddingProvider
from ..db.multi_tenant_store import MultiTenantGraphmindStore, RuntimeTenantContext
from ..utils.logger import create_logger


class ContextualRegistry:
    def __init__(self, multi_tenant_store: MultiTenantGraphmindStore, embedding_provider: EmbeddingProvider, config: ContextGraphConfig) -> None:
        self._multi_tenant_store = multi_tenant_store
        self._embedding = embedding_provider
        self._config = config
        self._logger = create_logger(config.debug)
        self._discovery_mode: dict[str, bool] = {}

    def is_discovery_mode(self, runtime_context: RuntimeTenantContext | None = None) -> bool:
        store = self._multi_tenant_store.get_store_for_runtime(runtime_context)
        cache_key = store.graph_name

        if cache_key in self._discovery_mode:
            return self._discovery_mode[cache_key]

        count = store.count_traces()
        is_discovery = count == 0
        self._discovery_mode[cache_key] = is_discovery

        if is_discovery:
            self._logger.info(
                "Discovery Mode: no prior traces found for tenant %s, project %s",
                store.tenant,
                store.project,
            )
        return is_discovery

    def get_relevant_context(
        self,
        intent_description: str,
        runtime_context: RuntimeTenantContext | None = None,
    ) -> FormattedContext:
        store = self._multi_tenant_store.get_store_for_runtime(runtime_context)
        embedding = self._embedding.embed(intent_description)
        limit = self._config.vector_search_limit or DEFAULT_VECTOR_SEARCH_LIMIT

        past_traces = store.find_similar_traces(embedding, limit)
        rules = store.get_active_rules()
        anti_patterns = store.get_anti_patterns()
        skills = store.get_skills_by_project()
        schema = store.get_schema_overview()

        self._logger.debug(
            "Retrieved context for tenant %s: %d traces, %d rules, %d anti-patterns, %d skills, %d entity types",
            store.tenant,
            len(past_traces),
            len(rules),
            len(anti_patterns),
            len(skills),
            len(schema.node_labels),
        )

        return FormattedContext(
            past_traces=past_traces,
            rules=rules,
            anti_patterns=anti_patterns,
            skills=skills,
            schema=schema,
        )

    def record_decision(
        self,
        trace: DecisionTrace,
        runtime_context: RuntimeTenantContext | None = None,
    ) -> str:
        """Save a decision trace with auto-generated embeddings."""
        store = self._multi_tenant_store.get_store_for_runtime(runtime_context)

        trace_text = "\n".join([
            f"Intent: {trace.intent.description}",
            *[f"Constraint ({c.type}): {c.description}" for c in trace.constraints],
            f"Action: {trace.action.description}",
            f"Justification: {trace.justification.description}",
        ])

        trace_emb, intent_emb = self._embedding.embed_batch([trace_text, trace.intent.description])
        action_emb = self._embedding.embed(trace.action.description)

        constraint_embeddings = (
            self._embedding.embed_batch([c.description for c in trace.constraints])
            if trace.constraints else []
        )

        # Enrich trace with embeddings
        trace.embedding = trace_emb
        trace.intent.embedding = intent_emb
        trace.action.embedding = action_emb
        for i, c in enumerate(trace.constraints):
            if i < len(constraint_embeddings):
                c.embedding = constraint_embeddings[i]

        trace_id = store.save_decision_trace(trace)

        # Embed concepts
        if trace.concepts:
            try:
                concept_embeddings = self._embedding.embed_batch(trace.concepts)
                for i, name in enumerate(trace.concepts):
                    store.ensure_concept(name, embedding=concept_embeddings[i])
            except Exception as e:
                self._logger.debug("Failed to embed concepts: %s", e)

        # Link precedents
        self._link_precedents(trace_id, trace_emb, store)

        # Reset discovery mode cache for this tenant
        self._discovery_mode.pop(store.graph_name, None)

        self._logger.debug("Recorded decision trace: %s for tenant: %s", trace_id, store.tenant)
        return trace_id

    def _link_precedents(self, new_trace_id: str, embedding: list[float], store) -> None:
        threshold = self._config.similarity_threshold or DEFAULT_SIMILARITY_THRESHOLD
        similar = store.find_similar_traces(embedding, 5)
        for scored in similar:
            if scored.trace.id and scored.trace.id != new_trace_id and scored.similarity >= threshold:
                store.create_precedent_link(new_trace_id, scored.trace.id, scored.similarity)
